refactor: 인증 관련 코드 리팩토링

tstunas opened the pull request:
### PR 타입(하나 이상의 PR 타입을 선택)
- [ ] 기능 추가   
- [x] 기능 수정   
- [x] 기능 삭제   
- [x] 버그 수정   
- [ ] 의존성, 환경 변수, 빌드 관련 코드 업데이트
- [ ] 기타 수정

### 변경 사항
- resolve #105 
  - 샘플 유저 생성 API 제거 (더 이상 테스트용도로 필요없다고 판단)
  - passport-local을 사용하여 로그인 로직 단순화
  - 소셜 로그인에서 내보내던 Kakao서비스, Naver서비스를 안 내보내고, Social서비스 하나를 내보내는 방식으로 리팩토링
  - 블락유저 여부를 캐싱하는 전략을 변경 (우준호 튜터님의 피드백을 반영) - 블락을 하면 1시간동안 캐시에 블락 유저로 등록
  - jwt, bcrypt, cache manager 관련 코드를 별도의 서비스로 분리
  - 인증 서비스의 스테이터스 코드, 메시지 약간 수정
  - 테스트 코드 재작성
- 로그아웃 API가 리프레시 토큰을 제대로 못 없애고 있어서 올바르게 수정 ([5506eec](https://github.com/chalkak2023/Chalkak-Backend/pull/141/commits/5506eec25ead0e9f1d93d55a2cd01e582c2f8ebc) 참조)

### 테스트 결과
- 정상 (테스트 코드 + 클라이언트 테스트)

### 추가사항
- 별로 급한 pr은 아니니 충분히 리뷰하고 합칠 예정

tstunas committed:
refactor: 더 이상 필요없는 샘플 유저 생성 API 제거

tstunas committed:
refactor: passport-local을 사용한 이메일 패스워드 로그인 리펙토링

+ 가입된 이메일이 아니거나 패스워드가 다르다면 404가 아니라 401을 되돌려주도록 변경

tstunas committed:
refactor: 블락유저 여부를 캐싱하는 전략을 변경(DB를 안 이용하고 캐시만 이용)

레디스만 사용하여 실시간 블락여부만 체크하기 때문에, 로그인의 경우에는 가드가 아니라 DB의 유저테이블에 isBlock 값을 검증

tstunas committed:
chore: 인증 API 코드 중 unused import 제거

tstunas committed:
refactor: 액세스토큰 만드는 메소드의 인자의 타입을 User로 변경.

- user에 \!를 붙여서 반드시 존재한다고 선언하던 걸 없애고 type narrow로 타입에서 null을 제거하여 해결

tstunas committed:
refactor: 인증에 관련된 jwt 코드들을 하나의 서비스 파일로 분리

tstunas committed:
refactor: AuthService의 복잡도를 낮추기 위해 CacheManger 관련 코드를 별도의 서비스로 분리

+ signout에서 캐시를 userId 키로 제거하고 있는 어긋나고 있는 코드를 수정

tstunas committed:
refactor: 소셜 로그인 모듈에서 내보내는 서비스를 하나로 줄임

인터페이스를 단순화함. 또한 외부 Oauth2 요청에 대한 응답의 인터페이스를 작성하여 타입 검증을 강화

tstunas committed:
refactor: 값이 존재하지 않는 것을 _.isNil을 사용하여 리팩토링

tstunas committed:
refactor: 인증 서비스에서 bcrypt 관련을 하나의 서비스로 독립

tstunas committed:
refactor: 인증 서비스의 예외 스테이터스 코드, 메시지 수정, 추가

tstunas committed:
chore: verifyToken 생성하는 private 메소드의 이름을 createVerifyToken으로 수정

tstunas committed:
test: 인증 컨트롤러, 서비스의 테스트 코드를 test 폴더로 옮기고 재작성

tstunas committed:
fix: 인증번호 저장할 때 누락된 ttl 추가

tstunas committed:
test: AuthCacheService 유닛 테스트 코드 작성

tstunas committed:
test: AuthJwtService 유닛 테스트 작성

tstunas committed:
refactor: auth.dto.ts에서 as const 제거

tstunas committed:
fix: MailerConfigService에서 port가 MAILER_HOST 환경변수를 참고하고 있었음

tstunas commented:
메일 환경변수 PORT가 제대로 적용되면서 추가 세팅이 필요해짐. 조사 중.

tstunas commented:
**로컬에서 테스트할 때, MAILER_PORT 환경변수 값을 465에서 587로 수정할 것.**
465는 SSL(보안 연결)할 때 쓰이는 포트이며, 그런 환경이 로컬에서 세팅되어있지 않다면 587로 해야함.
(지금까지 로컬에서 메일 보낼 때 587이 쓰였음)
추후, 본 서버 환경에서 465 포트로 보안연결로 사용할 이메일 만들어야함. (기존 이메일은 587로 보안연결없이 보내서 노출되었을 가능성 있음)

tstunas committed:
refactor: import한 Cache를 CacheManager로 별칭 사용하는 걸 복구

cchoseonghun commented on src/auth/auth.service.ts:
```diff
@@ -37,92 +32,75 @@ export class AuthService {
         message: '인증번호를 요청하지 않았거나 만료되었습니다.',
       });
     }
     if (verifyToken != cachedVerifyToken) {
```
!== 를 쓰는걸 추천

cchoseonghun commented on src/auth/auth.service.ts:
```diff
@@ -37,92 +32,75 @@ export class AuthService {
     } catch (e) {
-      throw new BadRequestException({
-        message: '회원가입에 적절하지 않은 이메일과 패스워드입니다.',
+      throw new ConflictException({
```
localUserRepository.insert에서 에러가 나면 catch문으로 올텐데 발생한 에러가 꼭 ConflictException이진 않아보입니다. 음 저라면 500 에러가 맞을거 같기도 하고

tstunas committed:
chore: 서비스를 담은 폴더의 이름을 service에서 services로 복수형으로 변경

cchoseonghun commented on src/auth/auth.service.ts:
```diff
@@ -211,57 +190,53 @@ export class AuthService {
-
-    const id = info.id;
-    let nickname = provider === 'kakao' ? info.kakao_account.profile.nickname : info.nickname;
+    const { accessToken: socialAccessToken, providerUserId, username: _username } = await this.socialService.validateSocialUser(provider, body);
+    let username = _username;
```
const username 그대로 받고 중복된 username이 있을 경우 
뒤에 랜덤번호를 붙인 newUsername을 사용하는건 어떨까요? 
굳이 username을 유지하기 위해 username으로 받은 뒤 
_username으로 변환시키고 이걸 다시 let username에 넣어 쓰는거보단 좋을 것 같습니다.

tstunas replied:
```typescript
const socialUserInfo = await this.socialService.validateSocialUser(provider, body);
const providerUserId = socialUserInfo.providerUserId;
let username = socialUserInfo.username;
```
이런 식으로 고치겠습니다. newUsername을 만들면 나중에 유저 생성할 때 username 사용과 newUsername 사용을 분기시켜야할 것 같아서 const와 let으로 따로따로 만들었습니다.

cchoseonghun replied:
```
let { accessToken: socialAccessToken, providerUserId, username }
```
아니면 그냥 애초에 let을 쓰는건 어떨까요

tstunas replied:
그러는 것도 나쁘지 않겠네요.

cchoseonghun replied:
물론 지금 코드 그대로 써도 무방하기 때문에 수정 안하셔도 되지만 제 기준에는 코드가 난잡해 보이긴 하네요.
어떻게 하실지는 진님 재량에 맡기겠습니다.

tstunas replied:
> ```
> let { accessToken: socialAccessToken, providerUserId, username }
> ```
> 
> 아니면 그냥 애초에 let을 쓰는건 어떨까요

이거 대로 수정했습니다.

cchoseonghun commented on src/auth/auth.service.ts:
```diff
@@ -211,57 +190,53 @@ export class AuthService {
         });
       } catch (err) {
-        throw new BadRequestException({
+        throw new ConflictException({
```
이 경우에도 insert에 대해 실패하는 경우가 중복되서만 이라고 보기 어려울 것 같습니다. 오히려 위에서 중복 체크를 했으니 catch에서 잡힌다면 conflict가 아닌 에러가 나올 확률이 더 크지 않을까요?

cchoseonghun commented on src/auth/auth.service.ts:
```diff
@@ -211,57 +190,53 @@ export class AuthService {
       });
     }
-    if (user!.isBlock) {
+    if (_.isNil(user)) {
```
211줄에 같은 코드가 있는 것 같은데 확인 부탁드립니다.

tstunas replied:
위의 if 블록 안에서 user를 재할당하고 나서 유효성 검사를 하는 거였는데 블록 안에 넣어야 옳겠네요.

tstunas replied:
수정 완료

cchoseonghun commented on src/auth/auth.service.ts:
```diff
@@ -271,21 +246,11 @@ export class AuthService {
   }
 
   async refreshAccessToken(accessToken: string, refreshToken: string) {
```
refreshAccessToken이라는 메서드 이름 보단 더 명확하게 무슨 역할을 하는지 알 수 있어야 된다고 봅니다.

tstunas replied:
액세스 토큰을 재생성한다는 의미에서 그렇게 지었는데, 아무래도 별도로 리프레시 토큰이 있으니까 모호한 뜻이 된 거 같습니다. renewAccessToken으로 바꾸겠습니다.

tstunas replied:
수정 완료

cchoseonghun commented on src/auth/entities/user.entity.ts:
```diff
@@ -34,7 +34,7 @@ export class User {
 
   @Column('varchar', { nullable: true })
-  providerUserId: string | null;
+  providerUserId: string | number | null;
```
결국 DB엔 varchar 컬럼이던데 number를 추가한 이유가 있을까요?

tstunas replied:
카카오나 네이버 중에 number로 id를 주는 애가 있어서 vscode에서 타입 오류라고 해서 지정했던 것 같습니다. 그냥 number인 애를 String해서 넣고 number 타입을 삭제하겠습니다.

cchoseonghun replied:
그럴 필요 없어보입니다. 그냥 궁금해서 물어본거였습니다. 
자동으로 형변환해서 varchar로 넣어주는거기 때문에 최대한 활용하죠. 
타입 확인 후 string으로 변환해서 전달하는거보다 진님 방법이 효율적이라 생각됩니다.

tstunas replied:
그럼 다시 revert 하겠습니다.

cchoseonghun replied:
순수히 질문 드리는건 더 확실하게 전달할 필요성이 있겠네요 ㅋㅋ
수정이 워낙 빠르셔서 괜히 revert 까지 하게 해버렸군요

tstunas replied:
revert 완료

cchoseonghun commented on src/auth/guard/local/local.strategy.ts:
```diff
@@ -0,0 +1,15 @@
+    super({ usernameField: 'email' });
+  }
+  async validate(email: string, password: string) {
+    const user = this.authService.validateLocalUser(email, password);
+    return user;
```
user에 담았다가 리턴할 필요가 있을까요?

cchoseonghun commented on src/auth/service/auth.cache.service.ts:
```diff
@@ -0,0 +1,29 @@
+  constructor(@Inject(CACHE_MANAGER) protected readonly cacheManager: Cache) {}
+
+  async getVerifyToken(type: verifyTokenType, email: string) {
+    return await this.cacheManager.get<number>(email + `_${type}`);
```
```
`${email}_${type}`
```
도 가능할텐데 섞어 쓴 이유가 있을지 궁금합니다.

cchoseonghun commented on src/auth/service/auth.hash.service.ts:
```diff
@@ -0,0 +1,15 @@
+
+@Injectable()
+export class AuthHashService {
+  constructor() {}
```
생성자의 쓰임새가 없다면 생략해도 되겠네요.

tstunas committed:
refactor: auth.service.ts 피드백 수정

tstunas committed:
chore: refreshAccessToken을 renewAccessToken으로 이름 변경

tstunas committed:
fix: user entity에서 providerUserId의 타입에서 number 제거

tstunas committed:
refactor: 나머지 피드백 수정

- 불필요한 const user 생성 제거
- 문자열 +로 합성하는 것과 템플릿 문법을 혼용하는 코드를 템플릿 문법만 사용하여 수정
- 비어있는 생성자를 제거

tstunas committed:
test: 수정한 내용에 맞게 예외의 종류를 바꾸고, providerUserId에 string만 집어넣음

tstunas committed:
Revert "fix: user entity에서 providerUserId의 타입에서 number 제거"

This reverts commit 9e71b8a0740a5f176f2e813b7cb319ca052007a5.

muja-code reviewed (approved):
확인
